icaruscode/Decode/MCDecoderICARUSTPCwROI_module.cc

```diff
@@ -501,51 +501,59 @@ void MCDecoderICARUSTPCwROI::processSingleLabel(art::Event&
-          ChannelArrayPair chanArr;
-          for (const auto rawDigit : rawDigitVec) {
-	    // Declare a temporary digit holder and resize it if downsizing the waveform
-	    raw::RawDigit::ADCvector_t rawDataVec(dataSize);
-	    // Decompress data into local holder
-	    raw::Uncompress(rawDigit->ADCs(), rawDataVec, rawDigit->Compression());
-	    // Fill into the data structure
-            raw::ChannelID_t channel = rawDigit->Channel();
-            unsigned int planeIdx       = fChannelToBoardWirePlaneMap.find(channel)->second.second.second;
-	    icarus_signal_processing::VectorFloat boardDataVec(rawDataVec.cbegin(),rawDataVec.cend());
-	    chanArr.first.push_back(daq::INoiseFilter::ChannelPlanePair(channel,planeIdx));
-	    chanArr.second.push_back(boardDataVec);
-	  }
-          processSingleImage(clockData, chanArr, coherentNoiseGrouping, concurrentRawDigits, concurrentRawRawDigits, coherentRawDigits, concurrentROIs);
-	});
+	      
+            unsigned int board = channelToBoardItr->second.first;
+	        auto mapIter = boardToRawDigitMap.find(board);
+	      
+            if (mapIter != boardToRawDigitMap.end()) mapIter->second.push_back(&rawDigit);
+            else                                     boardToRawDigitMap.insert({board,std::vector<const raw::RawDigit*>{&rawDigit}});
+	    }
+
+	    //std::cout << "boardToRawDigitMap.size()=" << boardToRawDigitMap.size() << std::endl;
+	    //for (auto elem : boardToRawDigitMap) {
+	    tbb::parallel_for (static_cast<std::size_t>(0),boardToRawDigitMap.size(),[&](size_t& r) 
+        {
+	        auto elem = std::next(boardToRawDigitMap.begin(),r);
 
+	        std::vector<const raw::RawDigit*>& rawDigitVec = elem->second;
+
+	        // Sort (use a lambda to sort by channel id)
+	        std::sort(rawDigitVec.begin(),rawDigitVec.end(),[](const raw::RawDigit* left, const raw::RawDigit* right) {return left->Channel() < right->Channel();});
+
+            ChannelArrayPair chanArr;
+
+            for (const auto rawDigit : rawDigitVec) 
+            {
+	            // Declare a temporary digit holder and resize it if downsizing the waveform
+	            raw::RawDigit::ADCvector_t rawDataVec(dataSize);
+	            // Decompress data into local holder
+	            raw::Uncompress(rawDigit->ADCs(), rawDataVec, rawDigit->Compression());
+	            // Fill into the data structure
+                raw::ChannelID_t channel = rawDigit->Channel();
+                unsigned int planeIdx       = fChannelToBoardWirePlaneMap.find(channel)->second.second.second;
+	            icarus_signal_processing::VectorFloat boardDataVec(rawDataVec.cbegin(),rawDataVec.cend());
+	            chanArr.first.push_back(daq::INoiseFilter::ChannelPlanePair(channel,planeIdx));
+	            chanArr.second.push_back(boardDataVec);
+	        }
+              processSingleImage(clockData, chanArr, coherentNoiseGrouping, concurrentRawDigits, concurrentRawRawDigits, coherentRawDigits, concurrentROIs);
```

Comment:
I think this line (552) is spurious, as processSingleImage is already called correctly below